File: devswitch/detect.py
# -*- coding: utf-8 -*-
from __future__ import print_function, unicode_literals

import glob
import os
import re
import shutil
import subprocess
from pathlib import Path
from typing import Iterable, List, Optional, Set

from . import paths
from .models import Runtime, java_label, major_of, node_label, parse_java_version, parse_node_version


def _run(cmd, timeout=4):
    # type: (List[str], int) -> str
    kwargs = {}
    if paths.IS_WINDOWS:
        # Keep console-less invocations (GUI, scanning) from flashing a window.
        kwargs["creationflags"] = getattr(subprocess, "CREATE_NO_WINDOW", 0x08000000)
    try:
        proc = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            timeout=timeout,
            check=False,
            **kwargs
        )
        out = proc.stdout or b""
        return out.decode("utf-8", "replace")
    except (OSError, subprocess.TimeoutExpired):
        return ""


def _unique_paths(values):
    # type: (Iterable[object]) -> List[Path]
    seen = set()  # type: Set[str]
    result = []
    for value in values:
        if not value:
            continue
        path = Path(os.path.expanduser(str(value)))
        try:
            resolved = str(path.resolve())
        except OSError:
            resolved = str(path)
        key = os.path.normcase(resolved) if paths.IS_WINDOWS else resolved
        if key in seen:
            continue
        seen.add(key)
        result.append(path)
    return result


def _which_all_windows(name):
    # type: (str) -> List[Path]
    found = []
    extensions = [".exe", ".cmd", ".bat"]
    for directory in os.environ.get("PATH", "").split(os.pathsep):
        directory = directory.strip('"')
        if not directory:
            continue
        base = Path(directory) / name
        for ext in extensions:
            candidate = Path(str(base) + ext)
            if candidate.is_file():
                found.append(candidate)
                break
    return found


def _which_all(name):
    # type: (str) -> List[Path]
    if paths.IS_WINDOWS:
        return _unique_paths(_which_all_windows(name))
    found = []
    which = shutil.which(name)
    if which:
        found.append(Path(which))
    try:
        proc = subprocess.run(
            ["bash", "-lc", "type -a -p {}".format(name)],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            timeout=3,
            check=False,
        )
        for line in (proc.stdout or b"").decode("utf-8", "replace").splitlines():
            line = line.strip()
            if line:
                found.append(Path(line))
    except (OSError, subprocess.TimeoutExpired):
        pass
    return _unique_paths(found)


def _is_shim(path):
    # type: (Path) -> bool
    try:
        if not path.is_file() or path.is_symlink():
            # Our shims are regular files; old user links are symlinks.
            if path.is_file() and not path.is_symlink():
                text = path.read_text(encoding="utf-8", errors="ignore")
                return "DevSwitch shim" in text
            return False
        text = path.read_text(encoding="utf-8", errors="ignore")
        return "DevSwitch shim" in text
    except (OSError, UnicodeError):
        return False


def _node_binary_names():
    # type: () -> tuple
    return ("node.exe",) if paths.IS_WINDOWS else ("node",)


def node_home_from_binary(binary):
    # type: (Path) -> Optional[Path]
    binary = Path(binary)
    try:
        binary = binary.resolve()
    except OSError:
        return None
    if binary.name not in _node_binary_names() or not binary.is_file():
        return None
    parent = binary.parent
    if parent.name == "bin":
        return parent.parent
    return parent


def java_home_from_binary(binary):
    # type: (Path) -> Optional[Path]
    binary = Path(binary)
    try:
        binary = binary.resolve()
    except OSError:
        return None
    names = ("java.exe",) if paths.IS_WINDOWS else ("java",)
    if binary.name not in names:
        return None
    if not paths.IS_WINDOWS and not os.access(str(binary), os.X_OK):
        return None
    parent = binary.parent
    if parent.name == "bin" and parent.parent.name == "jre":
        return parent.parent.parent
    if parent.name == "bin":
        return parent.parent
    return parent


def _inspect_node_binary(binary, home):
    # type: (Path, Path) -> Optional[Runtime]
    output = _run([str(binary), "-v"])
    version = parse_node_version(output)
    if not version:
        return None
    source = "system" if _looks_system(home) else "local"
    return Runtime(
        tool="node",
        version=version,
        major=major_of("node", version),
        home=str(home),
        binary=str(binary),
        source=source,
        label=node_label(version),
        vendor="Node.js",
    )


def inspect_node(home):
    # type: (Path) -> Optional[Runtime]
    home = Path(home)
    if paths.IS_WINDOWS:
        binary = home / "node.exe"
        if not binary.is_file():
            return None
        return _inspect_node_binary(binary, home)
    binary = home / "bin" / "node"
    if not binary.is_file():
        return None
    try:
        binary = binary.resolve()
        home = binary.parent.parent
    except OSError:
        return None
    return _inspect_node_binary(binary, home)


def _looks_system(home):
    # type: (Path) -> bool
    text = str(home)
    if paths.IS_WINDOWS:
        # 自己转小写而非依赖 os.path.normcase——它在非 Windows 宿主上是
        # 恒等函数，会让这条分支在跨平台测试里大小写不匹配。
        lower = text.replace("/", "\\").lower()
        if "\\program files\\" in lower or "\\program files (x86)\\" in lower:
            return True
        if lower.endswith("\\program files") or lower.endswith("\\program files (x86)"):
            return True
        for var in ("ProgramFiles", "ProgramFiles(x86)", "ProgramW6432"):
            base = os.environ.get(var)
            if base and lower.startswith(base.replace("/", "\\").lower() + "\\"):
                return True
        windir = os.environ.get("SystemRoot") or r"C:\Windows"
        return lower.startswith(windir.replace("/", "\\").lower() + "\\")
    return text.startswith("/usr")


def inspect_java(home):
    # type: (Path) -> Optional[Runtime]
    home = Path(home)
    exe = "java.exe" if paths.IS_WINDOWS else "java"
    binary = home / "bin" / exe
    if not binary.is_file():
        binary = home / "jre" / "bin" / exe
    if not binary.is_file():
        return None
    try:
        binary = binary.resolve()
        resolved_home = java_home_from_binary(binary) or home
    except OSError:
        return None
    output = _run([str(binary), "-version"])
    version = parse_java_version(output)
    if not version:
        return None
    vendor = "OpenJDK"
    if re.search(r"\bTemurin\b", output, re.I):
        vendor = "Temurin"
    elif re.search(r"\bZulu\b", output, re.I):
        vendor = "Zulu"
    elif re.search(r"\bGraalVM\b", output, re.I):
        vendor = "GraalVM"
    elif re.search(r"\bMicrosoft\b", output, re.I):
        vendor = "Microsoft"
    elif re.search(r"\bCorretto\b", output, re.I):
        vendor = "Corretto"
    elif re.search(r"\bLiberica\b", output, re.I):
        vendor = "Liberica"
    source = "system" if _looks_system(resolved_home) else "local"
    return Runtime(
        tool="java",
        version=version,
        major=major_of("java", version),
        home=str(resolved_home),
        binary=str(binary),
        source=source,
        label=java_label(version, vendor),
        vendor=vendor,
    )


def _glob_all(patterns):
    # type: (List[str]) -> List[Path]
    found = []
    for pattern in patterns:
        for item in glob.glob(pattern):
            found.append(Path(item))
    return found


def _candidate_node_homes():
    # type: () -> List[Path]
    home = paths.home()
    if paths.IS_WINDOWS:
        program_files = [
            os.environ.get(var)
            for var in ("ProgramFiles", "ProgramFiles(x86)", "ProgramW6432")
        ]
        patterns = []
        for base in program_files:
            if base:
                patterns.append(os.path.join(base, "nodejs"))
        nvm_root = os.environ.get("NVM_HOME") or os.path.join(
            str(home), "AppData", "Roaming", "nvm"
        )
        patterns.append(os.path.join(nvm_root, "v*"))
        patterns.append(str(home / "node-v*-win-*"))
        patterns.append(str(home / "scoop" / "apps" / "nodejs" / "*"))
        patterns.append(str(home / ".local" / "share" / "node-v*-win-*"))
        found = _glob_all(patterns)
    else:
        patterns = [
            str(home / "node-v*-linux-*"),
            str(home / ".local" / "share" / "node-v*-linux-*"),
            str(home / ".nvm" / "versions" / "node" / "*"),
            "/usr/local/n/versions/node/*",
            "/opt/node*",
            "/usr/local/lib/nodejs/*",
        ]
        found = _glob_all(patterns)
        usr_node = Path("/usr/bin/node")
        if usr_node.exists():
            found.append(Path("/usr"))
    for binary in _which_all("node"):
        if _is_shim(binary):
            continue
        derived = node_home_from_binary(binary)
        if derived is not None:
            found.append(derived)
    return _unique_paths(found)


def _candidate_java_homes():
    # type: () -> List[Path]
    home = paths.home()
    found = []  # type: List[Path]
    if paths.IS_WINDOWS:
        program_files = [
            os.environ.get(var)
            for var in ("ProgramFiles", "ProgramFiles(x86)", "ProgramW6432")
        ]
        patterns = []
        for base in program_files:
            if not base:
                continue
            for vendor_dir in (
                "Java",
                "Eclipse Adoptium",
                "Microsoft",
                "Zulu",
                "Amazon Corretto",
                "BellSoft",
            ):
                patterns.append(os.path.join(base, vendor_dir, "*"))
            # Android Studio ships a bundled JetBrains Runtime.
            patterns.append(os.path.join(base, "Android", "Android Studio", "jbr"))
        patterns.append(str(home / "jdk*"))
        patterns.append(str(home / "java*"))
        patterns.append(str(home / ".jdks" / "*"))
        found = _glob_all(patterns)
    else:
        for item in glob.glob("/usr/lib/jvm/*"):
            path = Path(item)
            name = path.name
            if name in ("default-java",) or name.startswith("openjdk-"):
                continue
            found.append(path)
        for pattern in [
            str(home / "jdk*"),
            str(home / "java*"),
            str(home / ".sdkman" / "candidates" / "java" / "*"),
            "/opt/jdk*",
            "/opt/java*",
            "/usr/java/*",
        ]:
            found.extend(_glob_all([pattern]))
    env_home = os.environ.get("JAVA_HOME")
    if env_home:
        found.append(Path(env_home))
    for binary in _which_all("java"):
        if _is_shim(binary):
            continue
        derived = java_home_from_binary(binary)
        if derived is not None:
            found.append(derived)
    return _unique_paths(found)


def scan_runtimes():
    # type: () -> List[Runtime]
    found = []
    seen = set()  # type: Set[str]
    for home in _candidate_node_homes():
        runtime = inspect_node(home)
        if runtime is None:
            continue
        key = "node:" + os.path.normcase(runtime.binary) if paths.IS_WINDOWS else "node:" + runtime.binary
        if key in seen:
            continue
        seen.add(key)
        found.append(runtime)
    for home in _candidate_java_homes():
        runtime = inspect_java(home)
        if runtime is None:
            continue
        key = "java:" + os.path.normcase(runtime.binary) if paths.IS_WINDOWS else "java:" + runtime.binary
        if key in seen:
            continue
        seen.add(key)
        found.append(runtime)
    found.sort(key=lambda item: (item.tool, item.version), reverse=True)
    return found


def infer_active(runtimes, tool):
    # type: (List[Runtime], str) -> Optional[Runtime]
    binary_name = "node.exe" if (tool == "node" and paths.IS_WINDOWS) else (
        "java.exe" if (tool == "java" and paths.IS_WINDOWS) else ("node" if tool == "node" else "java")
    )
    which = shutil.which(binary_name)
    if which:
        path = Path(which)
        if not _is_shim(path):
            try:
                resolved = str(path.resolve())
            except OSError:
                resolved = str(path)
            for runtime in runtimes:
                if runtime.tool == tool and runtime.binary == resolved:
                    return runtime
    if tool == "java":
        env_home = (os.environ.get("JAVA_HOME") or "").rstrip("/").rstrip("\\")
        for runtime in runtimes:
            home = runtime.home.rstrip("/").rstrip("\\")
            if runtime.tool == tool and (
                home == env_home or (paths.IS_WINDOWS and os.path.normcase(home) == os.path.normcase(env_home))
            ):
                return runtime
    items = [item for item in runtimes if item.tool == tool]
    return items[0] if items else None

File: internal/detect/detect.go
// Package detect 扫描本机运行时，平移自 devswitch/detect.py。
//
// 与 Python 版的实现差异：全程不执行子进程，版本信息一律来自
// 安装目录的静态元数据——
//   - Java：home/release 文件（JAVA_VERSION / IMPLEMENTOR，JDK 8+ 均有）
//   - Maven：lib/maven-core-*.jar 文件名
//   - Gradle：lib/ 下 gradle 核心 jar 文件名
//   - Node：目录名（node-vX.Y.Z-平台-架构），Windows 系统安装再查注册表卸载信息
// 优点是毫秒级完成（无需等待 JVM 启动）；代价是无元数据的罕见布局
// （如手工裸拷贝的二进制）不可识别，可用 devswitch import 手动补录。
package detect

import (
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"

	"github.com/xyztony999/devswitch/internal/models"
	"github.com/xyztony999/devswitch/internal/paths"
)

var isWindows = runtime.GOOS == "windows"

func normcase(s string) string {
	if isWindows {
		return strings.ToLower(s)
	}
	return s
}

func resolvePath(p string) string {
	abs, err := filepath.Abs(p)
	if err != nil {
		return p
	}
	real, err := filepath.EvalSymlinks(abs)
	if err != nil {
		return abs
	}
	return real
}

func uniquePaths(values []string) []string {
	seen := map[string]bool{}
	result := []string{}
	for _, value := range values {
		if value == "" {
			continue
		}
		key := normcase(resolvePath(value))
		if seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, value)
	}
	return result
}

// probeWhitelist：whichAll 只探测这些内部常量命令名。
var probeWhitelist = map[string]bool{
	"node": true, "java": true, "mvn": true, "gradle": true,
}

// whichAll 枚举 PATH 上的全部命中（纯目录扫描，无进程执行）。
func whichAll(name string) []string {
	if !probeWhitelist[name] {
		return nil
	}
	found := []string{}
	extensions := []string{""}
	if isWindows {
		extensions = []string{".exe", ".cmd", ".bat"}
	}
	for _, dir := range filepath.SplitList(os.Getenv("PATH")) {
		dir = strings.Trim(dir, `"`)
		if dir == "" {
			continue
		}
		for _, ext := range extensions {
			candidate := filepath.Join(dir, name+ext)
			if info, err := os.Stat(candidate); err == nil && !info.IsDir() {
				found = append(found, candidate)
				break
			}
		}
	}
	return uniquePaths(found)
}

func isShim(path string) bool {
	info, err := os.Lstat(path)
	if err != nil || info.IsDir() {
		return false
	}
	if info.Mode()&os.ModeSymlink != 0 {
		return false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	return strings.Contains(string(data), paths.ShimMarker)
}

// looksSystem 判断是否系统级安装（大小写自算，跨宿主语义一致）。
func looksSystem(home string) bool {
	if isWindows {
		lower := strings.ToLower(filepath.ToSlash(home))
		if strings.Contains(lower, "/program files/") || strings.Contains(lower, "/program files (x86)/") ||
			strings.HasSuffix(lower, "/program files") || strings.HasSuffix(lower, "/program files (x86)") {
			return true
		}
		for _, varName := range []string{"ProgramFiles", "ProgramFiles(x86)", "ProgramW6432"} {
			base := os.Getenv(varName)
			if base != "" && strings.HasPrefix(lower, strings.ToLower(filepath.ToSlash(base))+"/") {
				return true
			}
		}
		windir := os.Getenv("SystemRoot")
		if windir == "" {
			windir = `C:\Windows`
		}
		return strings.HasPrefix(lower, strings.ToLower(filepath.ToSlash(windir))+"/")
	}
	return strings.HasPrefix(home, "/usr")
}

func sourceOf(home string) string {
	if looksSystem(home) {
		return "system"
	}
	return "local"
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func fileExistsAny(pathsToTry ...string) bool {
	for _, p := range pathsToTry {
		if fileExists(p) {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// 元数据解析
// ---------------------------------------------------------------------------

var nodeDirVersionRe = regexp.MustCompile(`^node-v(\d+\.\d+\.\d+)-`)

// nodeVersionFromDirName 从 node-v22.23.3-win-x64 之类的目录名取版本。
func nodeVersionFromDirName(home string) string {
	m := nodeDirVersionRe.FindStringSubmatch(filepath.Base(filepath.Clean(home)))
	if m != nil {
		return m[1]
	}
	return ""
}

var releaseVersionRe = regexp.MustCompile(`(?m)^JAVA_VERSION="([^"]+)"`)
var releaseVendorRe = regexp.MustCompile(`(?m)^IMPLEMENTOR="([^"]+)"`)

// javaVersionFromRelease 读 home/release（JDK 8+ 标准文件）。
func javaVersionFromRelease(home string) (version string, vendor string) {
	data, err := os.ReadFile(filepath.Join(home, "release"))
	if err != nil {
		return "", ""
	}
	text := string(data)
	if m := releaseVersionRe.FindStringSubmatch(text); m != nil {
		version = m[1]
	}
	if m := releaseVendorRe.FindStringSubmatch(text); m != nil {
		vendor = m[1]
	}
	return version, vendor
}

var mavenCoreJarRe = regexp.MustCompile(`^maven-core-(\d+\.\d+(?:\.\d+)?(?:-[A-Za-z0-9.]+)?)\.jar$`)

// mavenVersionFromLibs 从 lib/maven-core-X.Y.Z.jar 取版本。
func mavenVersionFromLibs(home string) string {
	libDir := filepath.Join(home, "lib")
	entries, err := os.ReadDir(libDir)
	if err != nil {
		return ""
	}
	for _, entry := range entries {
		if m := mavenCoreJarRe.FindStringSubmatch(entry.Name()); m != nil {
			return m[1]
		}
	}
	return ""
}

var gradleCoreJarRe = regexp.MustCompile(`^gradle-(?:launcher|core|base|wrapper)-(?:\d+\.\d+(?:\.\d+)?(?:-[A-Za-z0-9.]+)?)[\.-].*\.jar$`)

// gradleVersionFromLibs 从 lib 下的核心 jar 文件名取版本。
func gradleVersionFromLibs(home string) string {
	libDir := filepath.Join(home, "lib")
	entries, err := os.ReadDir(libDir)
	if err != nil {
		return ""
	}
	for _, entry := range entries {
		name := entry.Name()
		if strings.HasPrefix(name, "gradle-") && strings.HasSuffix(name, ".jar") &&
			!strings.Contains(name, "api") && !strings.Contains(name, "internal") {
			// gradle-launcher-8.14.5.jar / gradle-wrapper-main-8.14.jar 等
			inner := strings.TrimSuffix(strings.TrimPrefix(name, "gradle-"), ".jar")
			parts := strings.SplitN(inner, "-", 2)
			if len(parts) == 2 && regexp.MustCompile(`^\d`).MatchString(parts[1]) {
				return parts[1]
			}
		}
	}
	return ""
}

func normalizeJavaVendor(vendor string) string {
	switch {
	case vendor == "":
		return "OpenJDK"
	case strings.Contains(strings.ToLower(vendor), "temurin"), strings.Contains(strings.ToLower(vendor), "eclipse"):
		return "Temurin"
	case strings.Contains(strings.ToLower(vendor), "microsoft"):
		return "Microsoft"
	case strings.Contains(strings.ToLower(vendor), "zulu"), strings.Contains(strings.ToLower(vendor), "azul"):
		return "Zulu"
	case strings.Contains(strings.ToLower(vendor), "corretto"), strings.Contains(strings.ToLower(vendor), "amazon"):
		return "Corretto"
	case strings.Contains(strings.ToLower(vendor), "liberica"), strings.Contains(strings.ToLower(vendor), "bellsoft"):
		return "Liberica"
	case strings.Contains(strings.ToLower(vendor), "graal"):
		return "GraalVM"
	case strings.Contains(strings.ToLower(vendor), "oracle"):
		return "Oracle"
	default:
		return vendor
	}
}

// ---------------------------------------------------------------------------
// 各工具 inspect
// ---------------------------------------------------------------------------

func nodeHomeFromBinary(binary string) (string, bool) {
	resolved := resolvePath(binary)
	name := filepath.Base(resolved)
	if isWindows {
		if name != "node.exe" {
			return "", false
		}
		return filepath.Dir(resolved), true
	}
	if name != "node" {
		return "", false
	}
	parent := filepath.Dir(resolved)
	if filepath.Base(parent) == "bin" {
		return filepath.Dir(parent), true
	}
	return parent, true
}

func inspectNode(home string) *models.Runtime {
	var binary string
	if isWindows {
		binary = filepath.Join(home, "node.exe")
	} else {
		binary = filepath.Join(home, "bin", "node")
	}
	if !fileExists(binary) {
		return nil
	}
	version := nodeVersionFromDirName(home)
	if version == "" && isWindows {
		version = nodeVersionFromRegistry(home)
	}
	if version == "" {
		return nil // 无元数据的布局（如 MSI 系统安装），可 devswitch import 补录
	}
	return &models.Runtime{
		Tool: models.Node, Version: version, Major: models.MajorOf(models.Node, version),
		Home: home, Binary: binary,
		Source: sourceOf(home), Label: models.NodeLabel(version), Vendor: "Node.js",
	}
}

func javaHomeFromBinary(binary string) (string, bool) {
	resolved := resolvePath(binary)
	want := "java"
	if isWindows {
		want = "java.exe"
	}
	if filepath.Base(resolved) != want {
		return "", false
	}
	parent := filepath.Dir(resolved)
	if filepath.Base(parent) == "jre" {
		return filepath.Dir(filepath.Dir(parent)), true
	}
	if filepath.Base(parent) == "bin" {
		return filepath.Dir(parent), true
	}
	return parent, true
}

func inspectJava(home string) *models.Runtime {
	exe := "java"
	if isWindows {
		exe = "java.exe"
	}
	binary := filepath.Join(home, "bin", exe)
	if !fileExists(binary) {
		binary = filepath.Join(home, "jre", "bin", exe)
		if !fileExists(binary) {
			return nil
		}
	}
	binary = resolvePath(binary)
	if resolvedHome, ok := javaHomeFromBinary(binary); ok {
		home = resolvedHome
	}
	version, vendor := javaVersionFromRelease(home)
	if version == "" {
		return nil
	}
	vendor = normalizeJavaVendor(vendor)
	return &models.Runtime{
		Tool: models.Java, Version: version, Major: models.MajorOf(models.Java, version),
		Home: home, Binary: binary,
		Source: sourceOf(home), Label: models.JavaLabel(version, vendor), Vendor: vendor,
	}
}

func inspectMaven(home string) *models.Runtime {
	binary := filepath.Join(home, "bin", "mvn.cmd")
	if !isWindows {
		binary = filepath.Join(home, "bin", "mvn")
	}
	if !fileExists(binary) {
		return nil
	}
	version := mavenVersionFromLibs(home)
	if version == "" {
		return nil
	}
	return &models.Runtime{
		Tool: models.Maven, Version: version, Major: models.MajorOf(models.Maven, version),
		Home: home, Binary: binary,
		Source: sourceOf(home), Label: "Apache Maven " + version, Vendor: "Apache",
	}
}

func inspectGradle(home string) *models.Runtime {
	binary := filepath.Join(home, "bin", "gradle.bat")
	if !isWindows {
		binary = filepath.Join(home, "bin", "gradle")
	}
	if !fileExists(binary) {
		return nil
	}
	version := gradleVersionFromLibs(home)
	if version == "" {
		return nil
	}
	return &models.Runtime{
		Tool: models.Gradle, Version: version, Major: models.MajorOf(models.Gradle, version),
		Home: home, Binary: binary,
		Source: sourceOf(home), Label: "Gradle " + version, Vendor: "Gradle",
	}
}

func Inspectors() map[models.Tool]func(string) *models.Runtime {
	return map[models.Tool]func(string) *models.Runtime{
		models.Node: inspectNode, models.Java: inspectJava,
		models.Maven: inspectMaven, models.Gradle: inspectGradle,
	}
}

// ---------------------------------------------------------------------------
// 候选目录
// ---------------------------------------------------------------------------

func globAll(patterns []string) []string {
	found := []string{}
	for _, pattern := range patterns {
		matches, _ := filepath.Glob(pattern)
		found = append(found, matches...)
	}
	return found
}

func programFilesVars() []string {
	return []string{
		os.Getenv("ProgramFiles"), os.Getenv("ProgramFiles(x86)"), os.Getenv("ProgramW6432"),
	}
}

func candidateNodeHomes() []string {
	home := paths.Home()
	var patterns []string
	if isWindows {
		for _, base := range programFilesVars() {
			if base != "" {
				patterns = append(patterns, filepath.Join(base, "nodejs"))
			}
		}
		nvmRoot := os.Getenv("NVM_HOME")
		if nvmRoot == "" {
			nvmRoot = filepath.Join(home, "AppData", "Roaming", "nvm")
		}
		patterns = append(patterns, filepath.Join(nvmRoot, "v*"))
		patterns = append(patterns,
			filepath.Join(home, "node-v*-win-*"),
			filepath.Join(home, "scoop", "apps", "nodejs", "*"),
			filepath.Join(home, ".local", "share", "node-v*-win-*"),
		)
	} else {
		patterns = append(patterns,
			filepath.Join(home, "node-v*-linux-*"),
			filepath.Join(home, ".local", "share", "node-v*-linux-*"),
			filepath.Join(home, ".nvm", "versions", "node", "*"),
			"/usr/local/n/versions/node/*",
			"/opt/node*",
			"/usr/local/lib/nodejs/*",
		)
	}
	found := globAll(patterns)
	for _, binary := range whichAll("node") {
		if isShim(binary) {
			continue
		}
		if home, ok := nodeHomeFromBinary(binary); ok {
			found = append(found, home)
		}
	}
	return uniquePaths(found)
}

func candidateJavaHomes() []string {
	home := paths.Home()
	found := []string{}
	if isWindows {
		for _, base := range programFilesVars() {
			if base == "" {
				continue
			}
			for _, vendorDir := range []string{"Java", "Eclipse Adoptium", "Microsoft", "Zulu", "Amazon Corretto", "BellSoft"} {
				found = append(found, globAll([]string{filepath.Join(base, vendorDir, "*")})...)
			}
			found = append(found, globAll([]string{filepath.Join(base, "Android", "Android Studio", "jbr")})...)
		}
		found = append(found, globAll([]string{
			filepath.Join(home, "jdk*"), filepath.Join(home, "java*"), filepath.Join(home, ".jdks", "*"),
		})...)
	} else {
		for _, item := range globAll([]string{"/usr/lib/jvm/*"}) {
			name := filepath.Base(item)
			if name == "default-java" || strings.HasPrefix(name, "openjdk-") {
				continue
			}
			found = append(found, item)
		}
		found = append(found, globAll([]string{
			filepath.Join(home, "jdk*"), filepath.Join(home, "java*"),
			filepath.Join(home, ".sdkman", "candidates", "java", "*"),
			"/opt/jdk*", "/opt/java*", "/usr/java/*",
		})...)
	}
	if envHome := os.Getenv("JAVA_HOME"); envHome != "" {
		found = append(found, envHome)
	}
	for _, binary := range whichAll("java") {
		if isShim(binary) {
			continue
		}
		if home, ok := javaHomeFromBinary(binary); ok {
			found = append(found, home)
		}
	}
	return uniquePaths(found)
}

func candidateBuildtoolHomes(tool models.Tool) []string {
	home := paths.Home()
	namePattern := "apache-maven-*"
	plain := "maven"
	if tool == models.Gradle {
		namePattern = "gradle-*"
		plain = "gradle"
	}
	var patterns []string
	if isWindows {
		patterns = append(patterns,
			filepath.Join(home, namePattern),
			filepath.Join(home, ".local", "share", namePattern),
			filepath.Join(home, "scoop", "apps", plain, "*"),
		)
		for _, base := range programFilesVars() {
			if base == "" {
				continue
			}
			if tool == models.Maven {
				patterns = append(patterns, filepath.Join(base, "apache-maven-*"))
			} else {
				patterns = append(patterns, filepath.Join(base, "*gradle*"))
			}
		}
	} else {
		optPattern := "/opt/apache-maven-*"
		if tool == models.Gradle {
			optPattern = "/opt/gradle-*"
		}
		patterns = append(patterns,
			filepath.Join(home, namePattern),
			filepath.Join(home, ".sdkman", "candidates", string(tool), "*"),
			optPattern,
			"/usr/share/"+string(tool)+"*",
		)
	}
	found := globAll(patterns)
	envVar := "MAVEN_HOME"
	cmd := "mvn"
	if tool == models.Gradle {
		envVar = "GRADLE_HOME"
		cmd = "gradle"
	}
	if envHome := os.Getenv(envVar); envHome != "" {
		found = append(found, envHome)
	}
	for _, binary := range whichAll(cmd) {
		if isShim(binary) {
			continue
		}
		found = append(found, filepath.Dir(filepath.Dir(binary)))
	}
	return uniquePaths(found)
}

// ---------------------------------------------------------------------------
// 扫描与推断
// ---------------------------------------------------------------------------

type probe struct {
	tool      models.Tool
	homes     func() []string
	inspector func(string) *models.Runtime
}

func ScanRuntimes() []models.Runtime {
	probes := []probe{
		{models.Node, candidateNodeHomes, inspectNode},
		{models.Java, candidateJavaHomes, inspectJava},
		{models.Maven, func() []string { return candidateBuildtoolHomes(models.Maven) }, inspectMaven},
		{models.Gradle, func() []string { return candidateBuildtoolHomes(models.Gradle) }, inspectGradle},
	}
	found := []models.Runtime{}
	seen := map[string]bool{}
	for _, p := range probes {
		for _, home := range p.homes() {
			runtime := p.inspector(home)
			if runtime == nil {
				continue
			}
			key := string(p.tool) + ":" + normcase(runtime.Binary)
			if seen[key] {
				continue
			}
			seen[key] = true
			found = append(found, *runtime)
		}
	}
	// 排序对齐 Python 版 (tool, version) 双键倒序
	for i := 1; i < len(found); i++ {
		for j := i; j > 0; j-- {
			a, b := found[j-1], found[j]
			if string(a.Tool) < string(b.Tool) ||
				(string(a.Tool) == string(b.Tool) && models.ParseVersionKey(a.Version).After(models.ParseVersionKey(b.Version))) {
				break
			}
			found[j-1], found[j] = found[j], found[j-1]
		}
	}
	return found
}

func InferActive(runtimes []models.Runtime, tool models.Tool) *models.Runtime {
	cmd := string(tool)
	if tool == models.Maven {
		cmd = "mvn"
	}
	for _, binary := range whichAll(cmd) {
		if isShim(binary) {
			continue
		}
		resolved := resolvePath(binary)
		for i := range runtimes {
			if runtimes[i].Tool == tool && runtimes[i].Binary == resolved {
				return &runtimes[i]
			}
		}
	}
	if tool == models.Java {
		envHome := strings.TrimRight(os.Getenv("JAVA_HOME"), "/\\")
		for i := range runtimes {
			home := strings.TrimRight(runtimes[i].Home, "/\\")
			if home == envHome || (isWindows && normcase(home) == normcase(envHome)) {
				return &runtimes[i]
			}
		}
	}
	for i := range runtimes {
		if runtimes[i].Tool == tool {
			return &runtimes[i]
		}
	}
	return nil
}
